pdns/ednsoptions.cc

```diff
@@ -23,6 +23,23 @@
 #include "ednsoptions.hh"
 #include "iputils.hh"
 
+bool getNextEDNSOption(const char* data, size_t dataLen, uint16_t& optionCode, uint16_t& optionLen)
+{
+  if (data == nullptr || dataLen < (sizeof(uint16_t) + sizeof(uint16_t))) {
+    return false;
+  }
+
+  size_t pos = 0;
+
+  optionCode = (static_cast<unsigned char>(data[pos]) * 256) + static_cast<unsigned char>(data[pos + 1]);
+  pos += EDNS_OPTION_CODE_SIZE;
+
+  optionLen = (static_cast<unsigned char>(data[pos]) * 256) + static_cast<unsigned char>(data[pos + 1]);
+  pos += EDNS_OPTION_LENGTH_SIZE;
```

Comment:
I prefer to have the conversion to uint16_t explicit:

```suggestion
  const uint8_t* p = reinterpret_cast<const uint8_t*>(data);
  optionCode = (static_cast<uint16_t>(p[pos]) * 256) + p[pos + 1];
  pos += EDNS_OPTION_CODE_SIZE;

  optionLen = (static_cast<uint16_t>(p[pos]) * 256) + p[pos + 1];
  pos += EDNS_OPTION_LENGTH_SIZE;

 ```